File: day18/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"

	"xorkevin.dev/gnom"
)

const (
	puzzleInput = "input.txt"
)

const (
	tokenKindDefault = iota
	tokenKindEOF
	tokenKindLparen
	tokenKindRparen
	tokenKindComma
	tokenKindNum
)

var (
	ErrParse = errors.New("Parse error")
)

type (
	Pair struct {
		val int
		lhs *Pair
		rhs *Pair
	}
)

func (p Pair) IsLiteral() bool {
	return p.lhs == nil
}

func (p Pair) buildString(b *strings.Builder) {
	if p.IsLiteral() {
		b.WriteString(strconv.Itoa(p.val))
		return
	}
	b.WriteByte('[')
	p.lhs.buildString(b)
	b.WriteByte(',')
	p.rhs.buildString(b)
	b.WriteByte(']')
}

func (p Pair) String() string {
	b := strings.Builder{}
	p.buildString(&b)
	return b.String()
}

func parsePairs(tokens []gnom.Token) (*Pair, []gnom.Token, error) {
	if len(tokens) == 0 {
		return nil, nil, ErrParse
	}
	top := tokens[0]
	switch top.Kind() {
	case tokenKindNum:
		{
			num, err := strconv.Atoi(top.Val())
			if err != nil {
				return nil, nil, ErrParse
			}
			return &Pair{
				val: num,
			}, tokens[1:], nil
		}
	case tokenKindLparen:
		{
			var lhs *Pair
			var err error
			lhs, tokens, err = parsePairs(tokens[1:])
			if err != nil {
				return nil, nil, err
			}
			if len(tokens) == 0 {
				return nil, nil, ErrParse
			}
			if tokens[0].Kind() != tokenKindComma {
				return nil, nil, ErrParse
			}
			var rhs *Pair
			rhs, tokens, err = parsePairs(tokens[1:])
			if err != nil {
				return nil, nil, ErrParse
			}
			if len(tokens) == 0 {
				return nil, nil, ErrParse
			}
			if tokens[0].Kind() != tokenKindRparen {
				return nil, nil, ErrParse
			}
			return &Pair{
				lhs: lhs,
				rhs: rhs,
			}, tokens[1:], nil
		}
	default:
		return nil, nil, ErrParse
	}
}

func (p *Pair) addLeft(v int) {
	if p.IsLiteral() {
		p.val += v
		return
	}
	p.lhs.addLeft(v)
}

func (p *Pair) addRight(v int) {
	if p.IsLiteral() {
		p.val += v
		return
	}
	p.rhs.addRight(v)
}

func (p *Pair) explode(depth int) (int, int, bool) {
	if p.IsLiteral() {
		return 0, 0, false
	}
	if depth > 3 {
		if !p.lhs.IsLiteral() || !p.rhs.IsLiteral() {
			log.Fatalln("Invalid reduction state")
		}
		l := p.lhs.val
		r := p.rhs.val
		p.val = 0
		p.lhs = nil
		p.rhs = nil
		return l, r, true
	}
	if l, r, ok := p.lhs.explode(depth + 1); ok {
		if r != 0 {
			p.rhs.addLeft(r)
		}
		return l, 0, true
	}
	if l, r, ok := p.rhs.explode(depth + 1); ok {
		if l != 0 {
			p.lhs.addRight(l)
		}
		return 0, r, true
	}
	return 0, 0, false
}

func (p *Pair) split() bool {
	if p.IsLiteral() {
		if p.val > 9 {
			l := p.val / 2
			r := p.val - l
			p.val = 0
			p.lhs = &Pair{
				val: l,
			}
			p.rhs = &Pair{
				val: r,
			}
			return true
		}
		return false
	}
	if ok := p.lhs.split(); ok {
		return true
	}
	return p.rhs.split()
}

func (p *Pair) reduceStep() bool {
	if _, _, ok := p.explode(0); ok {
		return true
	}
	return p.split()
}

func (p *Pair) Reduce() {
	for p.reduceStep() {
	}
}

func (p Pair) Magnitude() int {
	if p.IsLiteral() {
		return p.val
	}
	return 3*p.lhs.Magnitude() + 2*p.rhs.Magnitude()
}

func (p Pair) Clone() *Pair {
	if p.IsLiteral() {
		return &Pair{
			val: p.val,
		}
	}
	return &Pair{
		lhs: p.lhs.Clone(),
		rhs: p.rhs.Clone(),
	}
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	dfa := gnom.NewDfa(tokenKindDefault)
	dfaNum := gnom.NewDfa(tokenKindNum)
	dfa.AddDfa([]rune("0123456789"), dfaNum)
	dfaNum.AddDfa([]rune("0123456789"), dfaNum)
	dfa.AddPath([]rune("["), tokenKindLparen, tokenKindDefault)
	dfa.AddPath([]rune("]"), tokenKindRparen, tokenKindDefault)
	dfa.AddPath([]rune(","), tokenKindComma, tokenKindDefault)
	lexer := gnom.NewDfaLexer(dfa, tokenKindDefault, tokenKindEOF, map[int]struct{}{})

	var nums []*Pair
	var root *Pair
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		tokens, err := lexer.Tokenize([]rune(scanner.Text()))
		if err != nil {
			log.Fatal(err)
		}
		pair, _, err := parsePairs(tokens)
		nums = append(nums, pair.Clone())
		if root == nil {
			root = pair
		} else {
			root = &Pair{
				lhs: root,
				rhs: pair,
			}
		}
		root.Reduce()
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 1:", root.Magnitude())

	maxmag := 0
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			{
				k := &Pair{
					lhs: nums[i].Clone(),
					rhs: nums[j].Clone(),
				}
				k.Reduce()
				maxmag = max(maxmag, k.Magnitude())
			}
			{
				k := &Pair{
					lhs: nums[j].Clone(),
					rhs: nums[i].Clone(),
				}
				k.Reduce()
				maxmag = max(maxmag, k.Magnitude())
			}
		}
	}
	fmt.Println("Part 2:", maxmag)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: day18/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::iter::Peekable;
use std::slice::Iter;

const PUZZLEINPUT: &str = "input.txt";

#[derive(Debug)]
enum Token {
    Lparen,
    Rparen,
    Num(i32),
}

#[derive(Clone)]
enum Pair {
    Val(i32),
    Pair((Box<Pair>, Box<Pair>)),
}

impl Pair {
    fn add_left(&mut self, v: i32) {
        match self {
            Pair::Pair((lhs, _)) => lhs.add_left(v),
            Pair::Val(val) => *val += v,
        }
    }

    fn add_right(&mut self, v: i32) {
        match self {
            Pair::Pair((_, rhs)) => rhs.add_right(v),
            Pair::Val(val) => *val += v,
        }
    }

    fn explode(&mut self, depth: usize) -> Option<(i32, i32)> {
        let (lhs, rhs) = match self {
            Pair::Pair((lhs, rhs)) => (lhs, rhs),
            Pair::Val(_) => return None,
        };
        if depth > 3 {
            let lhs = match **lhs {
                Pair::Pair(_) => return None,
                Pair::Val(v) => v,
            };
            let rhs = match **rhs {
                Pair::Pair(_) => return None,
                Pair::Val(v) => v,
            };
            *self = Pair::Val(0);
            Some((lhs, rhs))
        } else {
            if let Some((l, r)) = lhs.explode(depth + 1) {
                if r != 0 {
                    rhs.add_left(r);
                }
                return Some((l, 0));
            }
            if let Some((l, r)) = rhs.explode(depth + 1) {
                if l != 0 {
                    lhs.add_right(l);
                }
                return Some((0, r));
            }
            None
        }
    }

    fn split(&mut self) -> bool {
        match self {
            Pair::Pair((lhs, rhs)) => lhs.split() || rhs.split(),
            Pair::Val(v) => {
                return if *v > 9 {
                    let l = *v / 2;
                    let r = *v - l;
                    *self = Pair::Pair((Box::new(Pair::Val(l)), Box::new(Pair::Val(r))));
                    true
                } else {
                    false
                }
            }
        }
    }

    fn reduce_step(&mut self) -> bool {
        if let Some(_) = self.explode(0) {
            return true;
        }
        return self.split();
    }

    fn reduce(&mut self) {
        while self.reduce_step() {}
    }

    fn magnitude(&self) -> i32 {
        match self {
            Pair::Pair((lhs, rhs)) => lhs.magnitude() * 3 + rhs.magnitude() * 2,
            Pair::Val(v) => *v,
        }
    }
}

fn tokenize(mut b: Peekable<Iter<u8>>) -> Result<Vec<Token>, Box<dyn std::error::Error>> {
    let mut tokens = Vec::new();
    while let Some(&c) = b.next() {
        match c {
            b'[' => tokens.push(Token::Lparen),
            b']' => tokens.push(Token::Rparen),
            b'0'..=b'9' => {
                let mut buf = vec![c];
                while let Some(&c2) = b.next_if(|&&i| i >= b'0' && i <= b'9') {
                    buf.push(c2);
                }
                tokens.push(Token::Num(String::from_utf8(buf)?.parse::<i32>()?));
            }
            _ => (),
        }
    }
    Ok(tokens)
}

fn parse_tokens(tokens: &[Token]) -> Option<(Pair, &[Token])> {
    let (head, rest) = match tokens {
        [head, rest @ ..] => (head, rest),
        &[] => return None,
    };
    match head {
        Token::Num(val) => Some((Pair::Val(*val), rest)),
        Token::Lparen => {
            let (lhs, rest1) = match parse_tokens(rest) {
                Some((k, rest)) => (k, rest),
                None => return None,
            };
            let (rhs, rest2) = match parse_tokens(rest1) {
                Some((k, rest)) => (k, rest),
                None => return None,
            };
            let rest3 = match rest2 {
                [Token::Rparen, rest @ ..] => rest,
                _ => return None,
            };
            Some((Pair::Pair((Box::new(lhs), Box::new(rhs))), rest3))
        }
        Token::Rparen => None,
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut root = None;
    let mut nums = Vec::new();
    for line in reader.lines() {
        let tokens = tokenize(line?.as_bytes().into_iter().peekable())?;
        let pair = match parse_tokens(&tokens) {
            Some((pair, _)) => pair,
            None => return Err("Invalid line".into()),
        };
        nums.push(pair.clone());
        let mut k = match root {
            Some(r) => Pair::Pair((Box::new(r), Box::new(pair))),
            None => pair,
        };
        k.reduce();
        root = Some(k);
    }
    match root {
        Some(r) => println!("Part 1: {}", r.magnitude()),
        None => return Err("No vals".into()),
    };

    match (0..nums.len() - 1)
        .into_iter()
        .flat_map(|i| {
            (i + 1..nums.len())
                .into_iter()
                .flat_map(move |j| vec![(i, j), (j, i)])
        })
        .map(|(i, j)| {
            let mut k = Pair::Pair((Box::new(nums[i].clone()), Box::new(nums[j].clone())));
            k.reduce();
            k.magnitude()
        })
        .max()
    {
        Some(v) => println!("Part 2: {}", v),
        None => return Err("No vals".into()),
    }

    Ok(())
}
